runtime/common/ArgumentConversion.cpp

```diff
@@ -100,6 +92,60 @@ static Value genConstant(OpBuilder &, cudaq::cc::StructType, void *,
 static Value genConstant(OpBuilder &, cudaq::cc::ArrayType, void *,
                          ModuleOp substMod, llvm::DataLayout &);
 
+static Value genConstant(OpBuilder &builder, const cudaq::state *v,
+                         ModuleOp substMod, llvm::DataLayout &layout,
+                         bool isSimulator) {
+  if (isSimulator) {
+    // The program is executed remotely, materialize the simulation data
+    // into an array and create a new state from it.
+    auto ctx = builder.getContext();
+    auto loc = builder.getUnknownLoc();
+    auto size = 1ULL << v->get_num_qubits();
+    auto is64Bit =
+        v->get_precision() == cudaq::SimulationState::precision::fp64;
+    auto eleTy = is64Bit ? ComplexType::get(Float64Type::get(ctx))
+                         : ComplexType::get(Float32Type::get(ctx));
+    auto arrTy = cudaq::cc::ArrayType::get(ctx, eleTy, size);
+
+    auto genConArray = [&]<typename T>() -> Value {
+      std::complex<T> data[size];
+      for (std::size_t i = 0; i < size; i++) {
+        data[i] = (*v)({i}, 0);
+      }
+      return genConstant(builder, arrTy, data, substMod, layout);
+    };
+
+    auto conArr = is64Bit ? genConArray.template operator()<double>()
+                          : genConArray.template operator()<float>();
+
+    cudaq::IRBuilder irBuilder(ctx);
+    auto createState = is64Bit ? cudaq::createCudaqStateFromDataFP64
+                               : cudaq::createCudaqStateFromDataFP32;
+    auto result = irBuilder.loadIntrinsic(substMod, createState);
+    assert(succeeded(result) && "loading intrinsic should never fail");
+
+    auto arrSize = builder.create<arith::ConstantIntOp>(loc, size, 64);
+    auto stateTy = cudaq::cc::StateType::get(ctx);
+    auto statePtrTy = cudaq::cc::PointerType::get(stateTy);
+    auto i8PtrTy = cudaq::cc::PointerType::get(builder.getI8Type());
+    auto buffer = builder.create<cudaq::cc::AllocaOp>(loc, arrTy);
+    builder.create<cudaq::cc::StoreOp>(loc, conArr, buffer);
+
+    auto cast = builder.create<cudaq::cc::CastOp>(loc, i8PtrTy, buffer);
+    auto statePtr = builder
+                        .create<func::CallOp>(loc, statePtrTy, createState,
+                                              ValueRange{cast, arrSize})
+                        .getResult(0);
+
+    // TODO: Delete the new state before function exit.
```

Comment:
@schweitzpgi what would be the best way to handle cleanup of the substitution? I'd like to add a function call at the function exit that deletes the new state.

Reply:
Discussed offline - will add another pass to:
- replace state* by the data in `cudaq.init_state` instruction
- remove `getNumberOfQubits(state*)` calls (replace by constant)
- if the state is no longer used, remove the `createFromData` call
- otherwise, add a call to delete the state before every exit from the function